### PR TITLE
fix(youtube/general-ads): restore swipe back to exit gesture

### PR DESCRIPTION
**Pull request by 0xrxL**
https://github.com/revanced/revanced-integrations/pull/264

Ok, i want to try to do it:

After the recent closed PR ```Double Back To Close``` by inotia00, I thought I'd reopen a new revised one without the double click code.

So...this is the result.

Closes: https://github.com/revanced/revanced-patches/issues/615

### CODE REVIEW

File: src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexit/fingerprints/RecyclerViewTopScrollingFingerprint.kt [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+object RecyclerViewTopScrollingFingerprint : MethodFingerprint(
+    opcodes = listOf(
+        Opcode.INVOKE_INTERFACE,
+        Opcode.MOVE_RESULT_OBJECT,
+        Opcode.CHECK_CAST,
+        Opcode.CONST_4,
+        Opcode.INVOKE_VIRTUAL,
+        Opcode.GOTO,
+        Opcode.IGET_OBJECT,
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexit/fingerprints/RecyclerViewTopScrollingFingerprint.kt:14]: This pattern is likely not gonna be unique. Strengthen the fingerprint without making it break easily.

File: src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexit/patch/BackToExitPatch.kt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import app.revanced.patcher.patch.BytecodePatch
+import app.revanced.patcher.patch.PatchResult
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patches.youtube.misc.fix.backtoexit.annotation.BackToExitCompatibility
+import app.revanced.patches.youtube.misc.fix.backtoexit.fingerprints.OnBackPressedFingerprint
+import app.revanced.patches.youtube.misc.fix.backtoexit.fingerprints.RecyclerViewScrollingFingerprint
+import app.revanced.patches.youtube.misc.fix.backtoexit.fingerprints.RecyclerViewTopScrollingParentFingerprint
+import app.revanced.patches.youtube.misc.fix.backtoexit.fingerprints.RecyclerViewTopScrollingFingerprint
+
+@Description("Close the app by tapping the back button from the home feed.")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexit/patch/BackToExitPatch.kt:17]: The description does not make fully sense, please explain it better.

```suggestion
@Description("Closes the app by tapping the back button from the home feed.")
```

File: src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexit/patch/BackToExitPatch.kt [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+@BackToExitCompatibility
+@Version("0.0.1")
+class BackToExitPatch : BytecodePatch(
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexit/patch/BackToExitPatch.kt:20]: The patch should not be named this way, instead it should be named what it does: `FixBackToExitGesturePatch`.

File: src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexit/patch/BackToExitPatch.kt [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    listOf(
+        OnBackPressedFingerprint,
+        RecyclerViewScrollingFingerprint,
+        RecyclerViewTopScrollingParentFingerprint,
+    )
+) {
+    override fun execute(context: BytecodeContext): PatchResult {
+        with (RecyclerViewScrollingFingerprint.result!!) {
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexit/patch/BackToExitPatch.kt:28]: The scope function `with` is abused in this patch, please remove it, its not used for its intended use case.

File: src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexit/patch/BackToExitPatch.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                scanResult.patternScanResult!!.endIndex,
+                invokeString("p0", "exitOnBackPressed", "Lcom/google/android/apps/youtube/app/watchwhile/WatchWhileActivity;")
+            )
+        }
+
+        return PatchResultSuccess()
+    }
+
+    private companion object {
+        fun invokeString(
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexit/patch/BackToExitPatch.kt:55]: This method should not simply build a string. It should be called `injectCall` and accept a method to add the instruction to, index to add the instruction, and the remaining relevant current parameters via a class.

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

**Comment by oSumAtrIX** [comment]
Please follow the standards in commits when opening PRs.

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]
The method `injectCall` still does not follow [this request](https://github.com/revanced/revanced-patches/pull/1405#discussion_r1059574290). For that reason, a lot of boilerplate duplicate code is left over in the  `execute` method.

File: src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        val recyclerViewScrollingFingerprint = RecyclerViewScrollingFingerprint.result!!
+        recyclerViewScrollingFingerprint.mutableMethod.addInstruction(
+            recyclerViewScrollingFingerprint.scanResult.patternScanResult!!.endIndex,
+            injectCall("", "onStartScrollView", "")
+        )
+
+        RecyclerViewTopScrollingFingerprint.resolve(context, RecyclerViewTopScrollingParentFingerprint.result!!.classDef)
+
+        val recyclerViewTopScrollingFingerprint = RecyclerViewTopScrollingFingerprint.result!!
+        println(recyclerViewTopScrollingFingerprint)
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt:37]: Please properly review your PR before you submit it.

```suggestion
```

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

File: src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        return PatchResultSuccess()
+    }
+
+    private companion object {
+        fun injectCall(
+            register: String,
+            methodName: String,
+            methodParams: String
+        ) =
+            "invoke-static {$register}, Lapp/revanced/integrations/patches/BackToExitPatch;->$methodName($methodParams)V"
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt:57]: The requested changes from [here](https://github.com/revanced/revanced-patches/pull/1405#discussion_r1059573974) have not been reflected to the integrations.

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

File: src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt (hunk 7 above)
Comment (0xrxL) [reply src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt:57]: > The method `injectCall` still does not follow [this request](https://github.com/revanced/revanced-patches/pull/1405#discussion_r1059574290). For that reason, a lot of boilerplate duplicate code is left over in the `execute` method.

Sorry, I misunderstood the request.

Fixed!

**Comment by 0xrxL** [comment]
> The method `injectCall` still does not follow [this request](https://github.com/revanced/revanced-patches/pull/1405#discussion_r1059574290). For that reason, a lot of boilerplate duplicate code is left over in the `execute` method.

Sorry, I misunderstood the request.

Fixed!

**Comment by oSumAtrIX** [comment]
No changes have been pushed @0xrxL.

File: src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        OnBackPressedFingerprint,
+        RecyclerViewScrollingFingerprint,
+        RecyclerViewTopScrollingParentFingerprint,
+    )
+) {
+    override fun execute(context: BytecodeContext): PatchResult {
+        val recyclerViewScrollingFingerprint = RecyclerViewScrollingFingerprint.result!!
+        injectCall(
+            recyclerViewScrollingFingerprint.mutableMethod,
+            recyclerViewScrollingFingerprint.scanResult.patternScanResult!!.endIndex,
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt:32]: This is duplicated three times. Why not make the signature of InjectCall accept this: `SomeFingerprint.injectCall(...)`

Comment (oSumAtrIX) [reply src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt:32]: Please also use a data class for the parameters register, method name, and method parameters as requested [here](https://github.com/revanced/revanced-patches/pull/1405#discussion_r1059574290).

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

**Comment by 0xrxL** [comment]
Ok, I removed a lot of duplicate code...and this work could be done better by using the "also" function, but in this case i'm not sure if it's useful for code readability.

**Comment by oSumAtrIX** [comment]
Pushed some changes, please review & make a final test so we can merge it.

File: src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+import app.revanced.patcher.patch.PatchResult
+import app.revanced.patcher.patch.PatchResultError
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patches.youtube.misc.fix.backtoexitgesture.annotation.FixBackToExitGestureCompatibility
+import app.revanced.patches.youtube.misc.fix.backtoexitgesture.fingerprints.OnBackPressedFingerprint
+import app.revanced.patches.youtube.misc.fix.backtoexitgesture.fingerprints.RecyclerViewScrollingFingerprint
+import app.revanced.patches.youtube.misc.fix.backtoexitgesture.fingerprints.RecyclerViewTopScrollingFingerprint
+import app.revanced.patches.youtube.misc.fix.backtoexitgesture.fingerprints.RecyclerViewTopScrollingParentFingerprint
+
+@Description("Closes the app by tapping the back button from the home feed.")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt:20]: [This comment](https://github.com/revanced/revanced-patches/pull/1405#discussion_r1059573815) has not been addressed yet.

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

File: src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt (hunk 9 above)
Comment (0xrxL) [reply src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt:20]: Oh...sorry. I only given a fast look to the highlighted word "Closes", and thought your request was referring just a typo problem. I'll fix the description.

Comment (0xrxL) [reply src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt]: Check if the new description is better.

File: src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+import app.revanced.patcher.patch.PatchResult
+import app.revanced.patcher.patch.PatchResultError
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patches.youtube.misc.fix.backtoexitgesture.annotation.FixBackToExitGestureCompatibility
+import app.revanced.patches.youtube.misc.fix.backtoexitgesture.fingerprints.OnBackPressedFingerprint
+import app.revanced.patches.youtube.misc.fix.backtoexitgesture.fingerprints.RecyclerViewScrollingFingerprint
+import app.revanced.patches.youtube.misc.fix.backtoexitgesture.fingerprints.RecyclerViewTopScrollingFingerprint
+import app.revanced.patches.youtube.misc.fix.backtoexitgesture.fingerprints.RecyclerViewTopScrollingParentFingerprint
+
+@Description("Fixes the broken 'Tap on Back to Exit' feature on the Homepage.")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/misc/fix/backtoexitgesture/patch/FixBackToExitGesturePatch.kt:20]:
```suggestion
@Description("Fixes the swipe back to exit gesture.")
```

**Review by oSumAtrIX — APPROVED** [review]
Please make a final test

**Comment by 0xrxL** [comment]
> Please make a final test

You broken the entire patch lol.

I'll try to fix it.

**Comment by Domiiniik** [comment]
LMAO

**Comment by 0xrxL** [comment]
@oSumAtrIX fingerprints are good now, but a patch is not applied, even though the fingerprint manages to be resolved.

I don't know how to fix it.

**Comment by 0xrxL** [comment]
This is the bytecode class not patched, despite the succesfully fingerprint resolution:

Immagine 2022-12-31 183509

**Comment by oSumAtrIX** [comment]
Are you sure the fingerprints correctly resolve? Also check the bytecode and not the decompiled result.

**Comment by 0xrxL** [comment]
> Are you sure the fingerprints correctly resolve? Also check the bytecode and not the decompiled result.

All fingerprint correctly resolved. I checked it with println() method.

After the apk patching I checked the result with jadx, and kak.smali appears to be not patched.

I really don't know what you broke.

**Comment by 0xrxL** [comment]
kak.smali -> e()

**Comment by oSumAtrIX** [comment]
breakpoint after it calls addInstruction and check if it added the correct instruction.

**Comment by 0xrxL** [comment]
> breakpoint after it calls addInstruction and check if it added the correct instruction.

Everything is good based on the debug, but the first patch is not applied in every case.

Immagine 2022-12-31 185134

**Comment by 0xrxL** [comment]
It seems like the addInstruction() method didn't work for an unknown reason.

**Comment by oSumAtrIX** [comment]
Are you sure you are opening the correct patched apk in JADX?

**Comment by 0xrxL** [comment]
> Are you sure you are opening the correct patched apk in JADX?

Yeah...is the usually file I open once patching is done.

**Comment by 0xrxL** [comment]
Anyway only the first patch is not applied. It's really weird.

**Comment by oSumAtrIX** [comment]
Breakpoint inside addInstruction then. Check what it does. My assumption is that the instruction string has a newline or something which makes addInstruction compile no instructions.

**Comment by 0xrxL** [comment]
> Breakpoint inside addInstruction then. Check what it does. My assumption is that the instruction string has a newline or something which makes addInstruction compile no instructions.

Ok, I think I understand why it doesn't work. Before your refactor the indexes were different, in the sense that they weren't all equal to endIndex, but there was also endIndex+1.

Indeed, when I switch to Smali tab of Jadx, the injectedCall appears:

Immagine 2022-12-31 190205

Was just wrongly placed...that's all.

**Comment by 0xrxL** [comment]
I already have a solution for this ;)

**Comment by oSumAtrIX** [comment]
Yep as assumed the compiler simply ignored decompiling it.

**Comment by 0xrxL** [comment]
> Yep as assumed before the compiler simply ignored decompiling it, the patch wasn't broken entirely by the refactor but the wrong fingerprints were used and the index.

Ok, now the patch works.

By the way...I have a question about patcher: why don't you add a way to check '.line x' too? Because, as you saw in the above screenshot, when there is a line '.line x' preceded by goto, problems like this occur.

**Comment by oSumAtrIX** [comment]
I don't follow? What do you mean?

**Comment by oSumAtrIX** [comment]
: only exist in smali for the assembler to create the correct offsets for offset instructions. They don't exist in Dalvik.

**Comment by 0xrxL** [comment]
> I don't follow? What do you mean?

This:

Immagine 2022-12-31 191354

How I should add a patch after a label, without incurring in patching problems?

Patcher can't detect a label like '.line x' or ':cond_x'

**Pull request merged**